src/lib/optimizer/strategy/exists_to_semijoin_rule.cpp

```diff
@@ -0,0 +1,187 @@
+      return LQPVisitation::VisitInputs;
+    }
+
+    // Now check if one side of the predicate is a column "outside" the subselect and the other inside
+    auto inner_column_expression = std::shared_ptr<LQPColumnExpression>();
+    auto parameter_expression = std::shared_ptr<ParameterExpression>();
+
+    if (subselect_predicate_expression->arguments[0]->type == ExpressionType::LQPColumn &&
+        subselect_predicate_expression->arguments[1]->type == ExpressionType::Parameter) {
+      // Column left, parameter right
+      inner_column_expression =
+          std::static_pointer_cast<LQPColumnExpression>(subselect_predicate_expression->arguments[0]);
+      parameter_expression =
+          std::static_pointer_cast<ParameterExpression>(subselect_predicate_expression->arguments[1]);
+    } else if (subselect_predicate_expression->arguments[0]->type == ExpressionType::Parameter &&
+               subselect_predicate_expression->arguments[1]->type == ExpressionType::LQPColumn) {
+      // Column right, parameter left
+      parameter_expression =
+          std::static_pointer_cast<ParameterExpression>(subselect_predicate_expression->arguments[0]);
+      inner_column_expression =
+          std::static_pointer_cast<LQPColumnExpression>(subselect_predicate_expression->arguments[1]);
+    } else {
+      // It's not - let's check the next predicate
+      return LQPVisitation::VisitInputs;
+    }
+
+    if (parameter_expression->parameter_expression_type != ParameterExpressionType::External) {
+      // Close, but not close enough. This is a parameter of the prepared statement type
+      return LQPVisitation::VisitInputs;
+    }
+
+    // Transform the parameter back into a column expression on the outside
+    auto parameter_id_iter = std::find(subselect->parameter_ids.cbegin(), subselect->parameter_ids.cend(),
+                                       parameter_expression->parameter_id);
+    auto outer_column_expression =
+        subselect->arguments[std::distance(subselect->parameter_ids.cbegin(), parameter_id_iter)];
+
+    // Build the join predicate
+    join_predicate = std::make_shared<BinaryPredicateExpression>(PredicateCondition::Equals, outer_column_expression,
+                                                                 inner_column_expression);
+
+    // We still need to visit the inputs to make sure that there is not more than one predicate
+    return LQPVisitation::DoNotVisitInputs;
+  });
+
+  if (!join_predicate) {
+    // We failed to identify the join predicate or there is more than one predicate
+    return _apply_to_inputs(node);
+  }
+
+  // Remove the predicate from the subselect (because it is now handled by the join) - if it is the top level node,
+  // we need to remove it by pointing the LQP to its input
+  if (subselect->lqp == subselect_predicate_node) {
+    subselect->lqp = subselect_predicate_node->left_input();
+  } else {
+    lqp_remove_node(subselect_predicate_node);
```

Comment:
Just leaving this here, I'll forget it otherwise: Clone the subselect LQP before doing this, some other node/expression might still reference it.

Reply:
Any good way to do this? When I call `subselect->lqp->deep_copy()`, I can't call `lqp_remove_node(subselect_predicate_node);`, because it still points to the original LQP. How can I get a pointer to the predicate node? Should `AbstractLQPNode::deep_copy` return a pair `(new_lqp, node_mapping)`?